File: sign_demo.py
#!/usr/bin/env python
import os.path
import ed25519


def main():
    keyfile = "keypair.dat"
    keyseed = "keyseed.dat"
    if not os.path.exists(keyfile):
        if os.path.exists(keyseed):
            print("Load key seed from", keyseed)
            with open(keyseed, "rb") as fp:
                seed = fp.read()
            if len(seed) != 32:
                print("Bad seed size, must be 32 bytes")
                return
            print("Create key from 32 bytes seed")
            sk = ed25519.SigningKey(seed)
            vk = sk.get_verifying_key()
        else:
            print("Generate new keypair...")
            sk, vk = ed25519.create_keypair()
        skb = sk.to_bytes()
        print("Save", len(skb), "bytes to", keyfile)
        skb = sk.to_bytes()
        with open(keyfile, "wb") as fp:
            fp.write(skb)
        seed = skb[:32]
        print("Save", len(seed), "bytes to", keyseed)
        with open(keyseed, "wb") as fp:
            fp.write(seed)
    else:
        print("Load key pair from", keyfile)
        keydata = open(keyfile, 'rb').read()
        if len(keydata) != 64:
            print("Bad key file size, must be 64 bytes")
            return
        sk = ed25519.SigningKey(keydata)
        vk = sk.get_verifying_key()

    print("--- Secret key ---")
    print("HEX:", sk.to_ascii(encoding="hex").decode())
    print("B64:", sk.to_ascii(encoding="base64").decode())
    print("--- Public key ---")
    print("HEX:", vk.to_ascii(encoding="hex").decode())
    print("B64:", vk.to_ascii(encoding="base64").decode())
    print("---")

    msgfile = "hello.msg"
    if not os.path.exists(msgfile):
        message = b"hello"
        with open(msgfile, "wb") as fp:
            fp.write(message)
    else:
        print("Load message from", msgfile)
        message = open(msgfile, "rb").read()

    print("Message to sign:", message)

    sigfile = "hello.sig"
    if not os.path.exists(sigfile):
        sig = sk.sign(message, encoding="base64")
        print("Sign message, signature in base64:")
        print(sig.decode())
        print("Save signature to", sigfile)
        with open(sigfile, "wt") as fp:
            fp.write(sig.decode())
    else:
        print("Load signature from", sigfile)
        sig = open(sigfile).read().encode()
        print("Signature of message in base64:")
        print(sig.decode())

    print("Verify signature")
    vk.verify(sig, message, encoding="base64")
    print("SUCCESS")


if __name__ == '__main__':
    main()

File: sign_demo.php
<?php

if (file_exists("keyseed.dat")) {
    echo "Load key seed from file\n";
    $sign_seed = file_get_contents("keyseed.dat");
} else if (file_exists("keypair.dat")) {
    echo "Load key pair from file\n";
    $sign_pair = file_get_contents("keypair.dat");
    $sign_seed = substr($sign_pair, 0, 32);
} else {
    echo "Generate key seed\n";
    $sign_seed = random_bytes(SODIUM_CRYPTO_SIGN_SEEDBYTES);
    echo "Save keyseed.dat\n";
    file_put_contents("keyseed.dat", $sign_seed);
}

$sign_pair = sodium_crypto_sign_seed_keypair($sign_seed);
$sign_secret = sodium_crypto_sign_secretkey($sign_pair);
$sign_public = sodium_crypto_sign_publickey($sign_pair);

if (!file_exists("keypair.dat")) {
    echo "Save keypair.dat\n";
    $keypair = substr($sign_pair, 0, 64);
    file_put_contents("keypair.dat", $keypair);
}

$sign_secret32 = substr($sign_secret, 0, 32);
echo "--- Private key ---\n";
echo "HEX: ".bin2hex($sign_secret32)."\n";
echo "B64: ".base64_encode($sign_secret32)."\n";
echo "--- Public key ---\n";
echo "HEX: ".bin2hex($sign_public)."\n";
echo "B64: ".base64_encode($sign_public)."\n";
echo "---\n";

if (!file_exists("hello.msg")) {
    $message = 'hello';
    echo "Save message to file hello.msg\n";
    file_put_contents("hello.msg", $message);
} else {
    echo "Load message from file hello.msg\n";
    $message = file_get_contents("hello.msg");
}

echo "Message to sign: ".$message."\n";

if (!file_exists("hello.sig")) {
    $signature = sodium_crypto_sign_detached($message, $sign_secret);
    $signature_encoded = base64_encode($signature);
    echo "Sinature of message:\n";
    echo $signature_encoded."\n";
    file_put_contents("hello.sig", $signature_encoded);
} else {
    echo "Load signature from file hello.sig\n";
    $signature_encoded = file_get_contents("hello.sig");
    echo $signature_encoded."\n";
    $signature = base64_decode($signature_encoded);
}

echo "Verify signature\n";
$res = sodium_crypto_sign_verify_detached($signature, $message, $sign_public);
var_dump($res); # true or false

?>
